### PR TITLE
Feature/Xline client auth

### PR DESCRIPTION
**Pull request by bsbds**
Implement auth client in xline-client

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline-client/src/clients/auth.rs [hunk 1]
```diff
@@ -41,4 +55,697 @@ impl AuthClient {
...
+        request: Req,
+    ) -> Result<Res> {
+        let propose_id = self.generate_propose_id();
+        let request = RequestWithToken::new_with_token(request.into(), self.token.clone());
+        let cmd = Command::new(vec![], request, propose_id);
+        let (cmd_res, sync_res) = self.curp_client.propose_indexed(cmd).await?;
+        let mut res_wrapper = cmd_res.decode();
+        res_wrapper.update_revision(sync_res.revision());
+        Ok(res_wrapper.into())
+    }
```
Comment (themanforfree) [inline xline-client/src/clients/auth.rs:316]: These two methods are very similar, I suggest merging them to reduce duplication

File: xline-client/src/clients/auth.rs [hunk 2]
```diff
@@ -41,4 +55,697 @@ impl AuthClient {
...
+        }
+    }
+}
+
+impl From<AuthenticateRequest> for xlineapi::AuthenticateRequest {
+    #[inline]
+    fn from(req: AuthenticateRequest) -> Self {
+        req.inner
+    }
+}
```
Comment (themanforfree) [inline xline-client/src/clients/auth.rs:359]: I suggest move all wrap types and the conversion to a separate file

File: xline-client/src/clients/auth.rs (hunk 1 above)
Comment (bsbds) [reply xline-client/src/clients/auth.rs]: done

File: xline-client/src/clients/auth.rs (hunk 2 above)
Comment (bsbds) [reply xline-client/src/clients/auth.rs]: done

File: xline-client/src/types/mod.rs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+/// Auth type definitions.
+pub mod auth;
```
Comment (themanforfree) [inline xline-client/src/types/mod.rs:2]: Plese move kv types here too

Comment (bsbds) [reply xline-client/src/types/mod.rs:2]: I'll do it in another PR

File: xline-client/tests/auth.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        AuthUserChangePasswordRequest, AuthUserDeleteRequest, AuthUserGetRequest,
+        AuthUserGrantRoleRequest, AuthUserRevokeRoleRequest, Permission,
+    },
+};
+use xlineapi::Type;
+
+mod common;
+
+#[tokio::test]
+async fn role() -> Result<()> {
```
Comment (Phoenix500526) [inline xline-client/tests/auth.rs:17]: A big test is not a good practice. Please divide it into serval small and concrete tests with meaningful names.

File: xline-client/tests/auth.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+    client
+        .role_delete(AuthRoleDeleteRequest::new(role2))
+        .await?;
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_user() -> Result<()> {
```
Comment (Phoenix500526) [inline xline-client/tests/auth.rs:111]: Same as above.

File: xline-client/tests/auth.rs (hunk 4 above)
Comment (bsbds) [reply xline-client/tests/auth.rs]: done

File: xline-client/src/lib.rs [hunk 6]
```diff
@@ -209,8 +209,25 @@ impl Client {
...
+                    );
+                    let resp = tmp_auth
+                        .authenticate(types::auth::AuthenticateRequest::new(username, password))
+                        .await?;
+
+                    Some(resp.token)
+                }
+                None => None,
+            }
+        };
```
Comment (iGxnon) [inline xline-client/src/lib.rs:230]: use `Option::map` here?

Comment (bsbds) [reply xline-client/src/lib.rs]: brackets removed. The authenticate is an async method so better keep this match syntax.

**Review by themanforfree — APPROVED** [review]

**Review by iGxnon — DISMISSED** [review]

**Review by themanforfree — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**